src/Common/src/TypeSystem/Common/TypeDesc.cs

```diff
@@ -103,9 +103,27 @@ public abstract partial class TypeDesc
 
         public override bool Equals(Object o)
         {
+            // Its only valid to compare two TypeDescs in the same context
+            Debug.Assert(o == null || !(o is TypeDesc) || Object.ReferenceEquals(((TypeDesc)o).Context, this.Context));
             return Object.ReferenceEquals(this, o);
         }
 
+#if DEBUG
+        public static bool operator ==(TypeDesc left, TypeDesc right)
```

Comment:
Would there be value in adding `operator==` and `operator!=` to `FieldDesc` and `MethodDesc`?

If you do that, you should be able to just use the operator in `MetadataVirtualMethodAlgorithm.cs` and get rid of the assert. Plus we get extra protection everywhere else we compare Field/MethodDesc.